pkg/distributor/field_detection_test.go

```diff
@@ -434,3 +435,133 @@ func Benchmark_optParseExtractLogLevelFromLogLineLogfmt(b *testing.B) {
+		detector := newFieldDetector(
+			validationContext{
+				discoverGenericFields:   make(map[string][]string, 0),
+				allowStructuredMetadata: true,
+			})
+		require.False(t, detector.shouldDiscoverGenericFields())
+	})
+	t.Run("disabled if structured metadata is not allowed", func(t *testing.T) {
+		detector := newFieldDetector(
+			validationContext{
+				discoverGenericFields:   map[string][]string{"trace_id": []string{"trace_id", "TRACE_ID"}},
+				allowStructuredMetadata: false,
+			})
+		require.False(t, detector.shouldDiscoverGenericFields())
+	})
+	t.Run("enabled if structured metadata is allowed and map is not empty", func(t *testing.T) {
+		detector := newFieldDetector(
+			validationContext{
+				discoverGenericFields:   map[string][]string{"trace_id": []string{"trace_id", "TRACE_ID"}},
+				allowStructuredMetadata: true,
+			})
+		require.True(t, detector.shouldDiscoverGenericFields())
+	})
+}
+
+func Test_DetectGenericFields(t *testing.T) {
+
+	detector := newFieldDetector(
+		validationContext{
+			discoverGenericFields: map[string][]string{
+				"trace_id": []string{"trace_id"},
+				"org_id":   []string{"org_id", "user_id", "tenant_id"},
+			},
+			allowStructuredMetadata: true,
+		})
+
+	for _, tc := range []struct {
+		name     string
+		labels   labels.Labels
+		entry    logproto.Entry
+		expected push.LabelsAdapter
+	}{
+		{
+			name: "no match",
+			labels: labels.Labels{
+				{Name: "env", Value: "prod"},
+			},
+			entry: push.Entry{
+				Line:               "log line does not match",
+				StructuredMetadata: push.LabelsAdapter{},
+			},
+			expected: push.LabelsAdapter{},
+		},
+		{
+			name: "stream label matches",
+			labels: labels.Labels{
+				{Name: "trace_id", Value: "8c5f2ecbade6f01d"},
+				{Name: "tenant_id", Value: "fake"},
+			},
+			entry: push.Entry{
+				Line:               "log line does not match",
+				StructuredMetadata: push.LabelsAdapter{},
+			},
+			expected: push.LabelsAdapter{
+				{Name: "trace_id", Value: "8c5f2ecbade6f01d"},
+				{Name: "org_id", Value: "fake"},
+			},
+		},
+		{
+			name: "metadata matches",
+			labels: labels.Labels{
+				{Name: "env", Value: "prod"},
+			},
+			entry: push.Entry{
+				Line: "log line does not match",
+				StructuredMetadata: push.LabelsAdapter{
+					{Name: "trace_id", Value: "8c5f2ecbade6f01d"},
+					{Name: "user_id", Value: "fake"},
+				},
+			},
+			expected: push.LabelsAdapter{
+				{Name: "trace_id", Value: "8c5f2ecbade6f01d"},
+				{Name: "org_id", Value: "fake"},
+			},
+		},
+		{
+			name: "logline (logfmt) matches",
+			labels: labels.Labels{
+				{Name: "env", Value: "prod"},
+			},
+			entry: push.Entry{
+				Line:               `msg="this log line matches" trace_id="8c5f2ecbade6f01d" org_id=fake duration=1h`,
+				StructuredMetadata: push.LabelsAdapter{},
+			},
+			expected: push.LabelsAdapter{
+				{Name: "trace_id", Value: "8c5f2ecbade6f01d"},
+				{Name: "org_id", Value: "fake"},
+			},
+		},
+		{
+			name: "logline (json) matches",
```

Comment:
what about multiple matches (ie. `org_id` and `tenant_id`)? a test documenting that behavior would be nice

Reply:
This was a good call! The behaviour was different between `json` and `logfmt` lines. `json` matches the first field from the config, `logfmt` matches the first matching field in the log line.

Reply:
Fixed with https://github.com/grafana/loki/pull/15188/commits/167f35013c7fa9532c43dcb2babc2709b009c451